javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/StandardEndpointLabels.qll

```diff
@@ -0,0 +1,125 @@
+/**
+ * For internal use only.
+ *
+ * Provides classes and predicates that are useful for endpoint filters.
+ *
+ * The standard use of this library is to make use of `isPotentialEffectiveSink/1`
+ */
+
+private import javascript
+private import semmle.javascript.filters.ClassifyFiles as ClassifyFiles
+private import semmle.javascript.heuristics.SyntacticHeuristics
+private import CoreKnowledge as CoreKnowledge
+private import StandardEndpointFilters as StandardEndpointFilters
+
+module Labels {
+  private newtype TEndpointLabel =
+    TLegacyEndpointLabel() or
+    TLegacyReasonSinkExcludedEndpointLabel(string innerReason) {
+      innerReason =
+        [
+          "argument to modeled function", //
+          "argument to sinkless library", //
+          "sanitizer", //
+          "predicate", //
+          "hash", //
+          "numeric", //
+          "in " + ["externs", "generated", "library", "test"] + " file" //
+        ]
+    } or
+    TLegacyModeledDbAccess() or
+    TLegacyModeledSink() or
+    TLegacyModeledStepSource() or
+    TLegacyModeledHttp(string kind) { kind = ["request", "response"] }
+
+  class EndpointLabel extends TEndpointLabel {
+    abstract string getLabel(DataFlow::Node n);
+
+    abstract string toString();
+  }
+
+  class LegacyEndpointLabel extends EndpointLabel, TLegacyEndpointLabel {
+    override string getLabel(DataFlow::Node n) {
+      n = StandardEndpointFilters::getALikelyExternalLibraryCall() and
+      result = "legacy/likely-external-library-call"
+      or
+      StandardEndpointFilters::flowsToArgumentOfLikelyExternalLibraryCall(n) and
+      result = "legacy/flows-to-argument-of-likely-external-library-call"
+    }
```

Comment:
This reads as more of a hierarchy of labels to me, and calling `getLabel` on something that is already a label doesn't entirely make sense to me.

Another way of organizing this would be to split up `LegacyEndpointLabel` into `LegacyExternalLibraryCallLabel` and `LegacyFlowsToArgumentOfLikelyExternalLibraryCallLabel` where the last two are subclasses of the first.  Then we could have a `Label#matchesEndpoint(DataFlow::Node n)` predicate.  Why use strings?  Is there a performance reason?  If we're leaving it to later, why not do it now?

Reply:
> calling getLabel on something that is already a label doesn't entirely make sense to me.

This is great feedback. I think the solution will be to rename the `EndpointLabel` to `EndpointLabeller`!

> Why use strings?

The reason I'm returning a string is that there's one instance per label-range (a class will only ever produce labels in a certain range like `legacy/reason-sink-excluded/**`, but that `**` may be a set that is open to extension. I want consuming code to be able to pick the labeller by matching on the *range* of the labeller by default as this will very likely be faster than filtering labels; but filtering on the label should be possible, because this is more fine-grained.

Example: consider a label that measures how large the function that surrounds an endpoint is. That'd have a range of, eg., `context/function/lines-of-code/**`, but specific labels returned by `getLabel` would be `context/function/lines-of-code/32`, `context/function/lines-of-code/4`, etc..

With the `matchesEndpoint` design, you'd not be able to select "all endpoints in functions that have 15 lines or less" in a natural manner (natural means: relying on labels alone, not by adding extra methods to that labeller and using them in the .ql files) because that labeller would need to have an infinite number of instances (which isn't possible).

I'll add a documentation commit that will make this design more clear 👍